### ISSUE TITLE
Buffering Error when reading and parsing packets

### ISSUE THREAD

**Issue by PeiwenChen** [OpenBMP/openbmp#17]
Hi,
I am feeding the `gobgp` BMP packets to `openbmpd`, and this is the logs shown when running `openbmpd`:
```
2016-07-26T20:07:14.960814 | DEBUG    |    parseBMP.cpp[00560] | parsePeerUpEventHdr | 172.xx.xxx.xx : Peer UP local address is IPv4 172.xx.xx.xx
2016-07-26T20:07:14.960820 | INFO     | ReadIncomingMsg    | 127.0.0.1: PEER UP Received, local addr=172.xx.xx.xx:24609 remote addr=172.xx.xx.xx:179
2016-07-26T20:07:14.960825 | DEBUG    |    parseBMP.cpp[00522] | bufferBMPMessage   | sock=29: Buffering 112 from socket
2016-07-26T20:07:14.960844 | WARN     | parseOpenMsg       | 172.xx.xx.xx: Could not read capabilities in open message due to buffer not containing the full param length
2016-07-26T20:07:14.960849 | ERROR    | handleUpEvent      | 172.xx.xx.xx: rtr=127.0.0.1: Failed to read sent open message
2016-07-26T20:07:14.960916 | INFO     | ReadIncomingMsg    | 127.0.0.1: Caught: Failed to read open message
2016-07-26T20:07:14.961015 | INFO     | resolveIp          | resolve: 127.0.0.1 to localhost
```

The first WARN is coming from:
https://github.com/OpenBMP/openbmp/blob/d4155c4bdee450c6afcb4e0b82c9aebf4a64c560/Server/src/bgp/OpenMsg.cpp#L97

It seems like buffering configuration issues? I tune some buffer setting in `openbmp.conf` but no good luck. Maybe I am not tuning it correctly.
Can you share any inputs?

**Comment by TimEvens** [comment]
It's not a buffer issue, but rather an issue with goBGP and the OPEN message.   It's likely goBGP is sending a malformed OPEN in the BMP PEER_UP.   I can update the code to allow that malformed update, but I would like to see what they are really sending.   Can you provide me the goBGP version/commit and goBGP config you are using?  I can then set it up here and test/validate.    My current goBGP works fine, so it's likely in a different version.  

You can also send me the packet capture of the RAW BMP feed.  You can get that using ```tcpdump -i any -s 0 -w <path_file for capture> tcp port 5000```  To send me the packet capture, you can email it to me at tim@openbmp.org or tievens@cisco.com.

**Comment by PeiwenChen** [comment]
@TimEvens 
Thanks for your very quick response and I already sent an email to "tim@openbmp.org".

**Comment by PeiwenChen** [comment]
@TimEvens 
I think I know why I received the `OPEN` packet now.
Because I am using gobgpd with BMP enabled, and I set the BMP server localhost, listen to 5555, the gobgpd sends both BGP packets and BMP packets to the port 5555.

Then the openbmpd listening to 5555 gets the BGP OPEN packet and failed to parse it.

**Comment by PeiwenChen** [comment]
@TimEvens 
Can we enable the logging of this:
https://github.com/OpenBMP/openbmp/blob/d4155c4bdee450c6afcb4e0b82c9aebf4a64c560/Server/src/bgp/OpenMsg.cpp#L84

It seems there is no option to enable this `SELF_DEBUG` in the `openbmpd.conf` file.

**Comment by TimEvens** [comment]
I wasn't able to reproduce it till I used your open message.  I found the issue in gobgp and how it's parsing the graceful restart capability.  In your case, Arista is sending graceful restart capability with a length of 2, which is fine.   goBGP is incorrectly parsing the capability, which results in it incorrectly parsing graceful restart as well as any other capability that follows it.  gobgp regenerates the OPEN message for the received OPEN in BMP (it does it also when you do ```gobgp neighbor -j```) .  It's then that it regenerates it incorrectly with incorrect information.   

 
You can work around this issue without an upgrade/change in gobgpd if you disable arista from sending the graceful restart capability.    If you can't get it to stop, then you should try out my fix that I submitted as a pull request. 

 
gobgp Issue: https://github.com/osrg/gobgp/issues/1037

**Comment by TimEvens** [comment]
goBGP pull request has been accepted and should be available soon.

**Comment by PeiwenChen** [comment]
Huge thanks @TimEvens  fixing this!

**Comment by PeiwenChen** [comment]
They shipped V1.10!
